crates/holochain_types/src/app/app_bundle.rs

```diff
@@ -330,10 +302,6 @@ pub enum CellProvisioningOp {
     ProvisionOnly(DnaFile, u32),
     /// The specified installed_hash does not match the actual hash of the DNA selected for provisioning. Expected: {0}, Actual: {1}
     HashMismatch(DnaHash, DnaHash),
-    /// Ambiguous result, needs manual resolution; can't provision (should this be an Err?)
-    Conflict(CellProvisioningConflict),
+    /// Ambiguous result, needs more information; can't provision (should this be an Err?)
+    Unresolvable(String),
```

Comment:
Should this be structured rather than a string? This is the case where we need the user to resolve between two or more options right?

Reply:
Yes, but the course of action would be to call [`GetCompatibleCells`](https://github.com/holochain/holochain/blob/293d6e775b3f02285b831626c9911802207a8d85/crates/holochain_conductor_api/src/admin_interface.rs#L389), so the error doesn't need to return that info

Reply:
@ThetaSinner I think your comment is actually mostly inspired by the doc for this variant, which is outdated from what we eventually decided on. Rather than having to resolve only ambiguities, we require the user to specify the CellId for every UseExisting role at install time. So, the user really already has all the info they need at install time, and if they fail to specify all the necessary info, this error can't give them any more info than they already had the first time they tried. It's just an implementation error on the app dev's part that needs to be fixed.

So I doubled down on not making this a semantic error. Have a look at this commit and see if you agree https://github.com/holochain/holochain/pull/4075/commits/6ca542834ae6ba1a2c2aaadb3abfbbe098de32c3

Reply:
Do they? I thought adding that dna hash resolution admin method made sense as an option but if you have to use it up front then the program doing the install needs to be able to compute `DnaHash`es. You certainly can't do that with `hc`, so this won't be usable by hand. The launcher can probably get hold of our Rust code and use it but I'm not certain what's publicly usable.

With this way of preparing the install, I also think we should have a test that does an install the way that it'll have to be done. The current test uses its knowledge of the hashes but it would be preferable to set up a test wasm that does the use existing. That way we can make sure the interface is coherent, even if it's not particularly concise.

Reply:
I'm not following. The flow is, you first call `GetCompatibleCells`, which returns `BTreeSet<(InstalledAppId, BTreeSet<CellId>)>`. From that info, whatever your runtime is, e.g. Launcher, has everything it needs to show human-readable info about which apps and which DNAs are compatible, to drive a UI experience of resolving ambiguities in that moment. Then you have to pass your existing cell choices in during InstallApp. The error we're talking about happens if you failed to make a choice. For every UseExisting role, you need to provide a choice, whether there is one or many possible choices, or else you will get this error. That's why I don't think the error needs to be sophisticated in any way, because it's always invalid to not supply values here regardless of conductor state.

Reply:
The new logic makes sense. I was just wondering what that flow actually looks like and now I'm calling out that I don't see a test that implements this flow. Am I missing that? I don't see a test that calls `GetCompatibleCells`

Reply:
@ThetaSinner [this test](https://github.com/holochain/holochain/blob/eb7d3b9f08610e57ac90073d5969a76b4bb2e743/crates/holochain/src/conductor/tests/install_app_bundle.rs#L425) calls the essential function, which is basically the implementation of `GetCompatibleCells`. I don't see it necessary to call the websocket method directly since we have plenty of other tests making sure that websocket calls return properly.